File: tobas_core/tobas_rqt_py/tobas_rqt_py/widgets/stacked_widget.py
from typing import override
from typing import Iterator
from PyQt5.QtWidgets import QWidget, QStackedWidget

from ..utils import update_event_loop


class StackedWidget(QStackedWidget):
    """
    ===== QStackedWidgetとの違い =====
    - イテレータを定義
    - setCurrentIndexを安定化
    - 追加メソッド
    """

    def __iter__(self) -> Iterator[QWidget]:
        for i in range(self.count()):
            yield self.widget(i)

    @override
    def setCurrentIndex(self, index: int) -> None:
        # インデックスを更新
        super().setCurrentIndex(index)

        # Qtのイベントループを更新
        update_event_loop()

        # 画面を更新
        self.update()

    def clear(self) -> None:
        """全てのウィジェットを削除し，メモリを開放する．"""
        while self.count() > 0:
            widget = self.widget(0)
            self.removeWidget(widget)
            widget.deleteLater()

File: tobas_gui/tobas_qt_tools/src/widgets/stacked_widget.cpp
#include <QCoreApplication>

#include "tobas_qt_tools/widgets/stacked_widget.hpp"

namespace qt
{
void StackedWidget::clear()
{
  while (count() > 0)
  {
    auto tar_widget = widget(0);
    removeWidget(tar_widget);
    tar_widget->deleteLater();
  }
}

void StackedWidget::setCurrentIndex(int index)
{
  // インデックスを更新
  super::setCurrentIndex(index);

  // Qtのイベントループを更新
  QCoreApplication::processEvents();

  // 画面を更新
  update();
}
}  // namespace qt
